lib/graph_metrics.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/python
+# Copyright 2014 Google.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Functions for computing graph-based metrics over scores."""
+
+import encoder
+import fileset_picker
+import math
+import numpy
+import optimizer
+import visual_metrics
+
+class Error(Exception):
+  pass
+
+class ScoreGroup(object):
+  """The scores describing a single codec's run results."""
+
+  def __init__(self, filename, codec, score_function):
+    self.name = codec.name
+    self.videofile = encoder.Videofile(filename)
+    self.fileset = optimizer.FileAndRateSet()
+    self.fileset.AddFilesAndRates([filename],
+        fileset_picker.ChooseRates(self.videofile.width,
+                                   self.videofile.framerate))
+    self.my_optimizer = optimizer.Optimizer(codec,
+                                            file_set = self.fileset,
+                                            score_function=score_function)
+    self.filename = filename
+    self.encoder = self.my_optimizer.BestOverallEncoder()
+    if not self.encoder:
+      raise Error('No overall encoder for %s on %s' % (codec.name, filename))
+    self.points = None
+
+  def dataPoints(self):
+    if not self.points:
+      self.points = []
+      for rate in self.fileset.AllRatesForFile(self.filename):
```

Comment:
Not sure if this would be an issue, but if self.fileset.AllRatesForFile(self.filename) evaluates to an empty iterator, then self.points would remain [] i.e. False, and the caching would not work, i.e. the if not self.points above would trigger each time this method is called.